api/src/client.rs

```diff
@@ -29,203 +30,220 @@ use tokio::runtime::Runtime;
 
 pub type ClientResponseFuture<T> = Box<dyn Future<Item = T, Error = Error> + Send>;
 
-/// Helper function to easily issue a HTTP GET request against a given URL that
-/// returns a JSON object. Handles request building, JSON deserialization and
-/// response code checking.
-pub fn get<'a, T>(url: &'a str, api_secret: Option<String>) -> Result<T, Error>
-where
-	for<'de> T: Deserialize<'de>,
-{
-	handle_request(build_request(url, "GET", api_secret, None)?)
+pub struct GrinClient {
+	pub api_secret: Option<String>,
+	pub proxy_addr: Option<String>,
+	client: Client<hyper_rustls::HttpsConnector<HttpConnector>, Body>,
 }
 
-/// Helper function to easily issue an async HTTP GET request against a given
-/// URL that returns a future. Handles request building, JSON deserialization
-/// and response code checking.
-pub fn get_async<'a, T>(url: &'a str, api_secret: Option<String>) -> ClientResponseFuture<T>
-where
-	for<'de> T: Deserialize<'de> + Send + 'static,
-{
-	match build_request(url, "GET", api_secret, None) {
-		Ok(req) => Box::new(handle_request_async(req)),
-		Err(e) => Box::new(err(e)),
+impl Default for GrinClient {
+	fn default() -> Self {
+		GrinClient::new(None, None)
 	}
 }
 
-/// Helper function to easily issue a HTTP GET request
-/// on a given URL that returns nothing. Handles request
-/// building and response code checking.
-pub fn get_no_ret(url: &str, api_secret: Option<String>) -> Result<(), Error> {
-	let req = build_request(url, "GET", api_secret, None)?;
-	send_request(req)?;
-	Ok(())
-}
+impl GrinClient {
+	pub fn new(api_secret: Option<String>, proxy_addr: Option<String>) -> GrinClient {
+		let https = hyper_rustls::HttpsConnector::new(1);
+		let client = Client::builder().build::<_, Body>(https);
+		GrinClient {
+			api_secret,
+			proxy_addr,
```

Comment:
Actually, one related question: this is not used currently. How are you meant to use it?

Reply:
Similar question/comment, this may require some extra plumbing using [hyper-proxy](https://lib.rs/crates/hyper-proxy). Is this PR just meant to lay the initial ground work?